datafusion/catalog-listing/src/table.rs

```diff
@@ -336,17 +336,64 @@ impl ListingTable {
         self.options.format.file_source(table_schema)
     }
 
-    /// If file_sort_order is specified, creates the appropriate physical expressions
+    /// Creates output ordering from user-specified file_sort_order or derives
+    /// from file orderings when user doesn't specify.
+    ///
+    /// If user specified `file_sort_order`, that takes precedence.
+    /// Otherwise, attempts to derive common ordering from file orderings in
+    /// the provided file groups.
     pub fn try_create_output_ordering(
         &self,
         execution_props: &ExecutionProps,
+        file_groups: &[FileGroup],
     ) -> datafusion_common::Result<Vec<LexOrdering>> {
-        create_lex_ordering(
-            &self.table_schema,
-            &self.options.file_sort_order,
-            execution_props,
-        )
+        // If user specified sort order, use that
+        if !self.options.file_sort_order.is_empty() {
+            return create_lex_ordering(
+                &self.table_schema,
+                &self.options.file_sort_order,
+                execution_props,
+            );
+        }
+
+        // Otherwise, try to derive from file orderings
+        Ok(derive_common_ordering_from_files(file_groups))
+    }
+}
+
+/// Derives a common ordering from file orderings across all file groups.
+///
+/// Returns the common ordering if all files have compatible orderings,
+/// otherwise returns an empty Vec (no ordering).
+fn derive_common_ordering_from_files(file_groups: &[FileGroup]) -> Vec<LexOrdering> {
+    // Collect all file orderings
+    let mut all_orderings: Vec<&LexOrdering> = Vec::new();
+    for group in file_groups {
+        for file in group.iter() {
+            if let Some(ordering) = &file.ordering {
+                all_orderings.push(ordering);
+            } else {
+                // If any file has no ordering, we can't derive a common ordering
```

Comment:
We may add some debug or tracing log here:

```suggestion
                // If any file has no ordering, we can't derive a common ordering
                tracing!(
        "Cannot derive common ordering: file {} has different ordering ({:?}) than first file ({:?})",
        file.object_meta.location, ordering, first
    );
```